### PR TITLE
Triangle Count with groupReduceOn* methods

### PR DESCRIPTION
**Pull request by andralungu**

### CODE REVIEW

File: src/main/java/example/TriangleCount.java [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+	}
+
+	// *************************************************************************
+	// UTIL METHODS
+	// *************************************************************************
+
+	private static boolean fileOutput = false;
+	private static String edgeInputPath = null;
+	private static String outputPath = null;
+	private static Integer maxItertations = TriangleCountData.MAX_ITERATIONS;
```
Comment (vasia) [inline src/main/java/example/TriangleCount.java]: you can get rid of the iterations arg :-)

**Comment by vasia** [comment]
I had something quite different in mind for this one.
Here's the idea:

In the "vertex-centric" approach, you have the following steps:
- each vertex sends a msg with its id to larger-id neighbors. This can be realized with a groupReduceOnNeighbors as you have already done. The difference is that you have to generate those "messages" for the higher-id neighbors, i.e. collect a pair of (vertexId, highNeighborId) for each of the higher-id neighbors. Using your example data, vertex 1 would collect the pairs (2, 1) and (3, 1).
  Then, you group these values by id and attach the resulting sets to the vertices. Using your example data again, vertex 3 will have the set [1, 2] as vertex value (the messages it received).
- Next, with a second groupReduceOnNeighbors, you simulate the second superstep, i.e. propagating each received message to all neighbors with higher id. Again, you will have to collect pairs of (neighborID, msg) to simulate the messages. Using the example data, vertex 3 will collect the pairs (4, 1), (5, 1), (6, 1), (4, 2), (5, 2) and (6, 2). Again, you group by ID and attach these values to the receiving vertices.
- Finally, you need to check for existing triangles. Each vertex which has received a message, will only have to check if the received "msg" is a neighbor ID, i.e. if the third edge exists. Using the example data, vertex 3 will discover triangle 1-2-3 (edge 3-1), vertex 6 will discover triangle 2-3-6 (edge 6-2) and vertex 5 will discover triangle 3-4-5 (edge 5-3). Notice that it's always the vertex with the highest ID which discovers a triangle.

I think it will be helpful to manually "run" this example with the example data on paper, to understand what is going on. At least visualization is always helpful to me :-)

**Comment by andralungu** [comment]
Hi @vasia ,

For this example, I wanted to act like a "rebel" and do my own implementation. You have to admit it still counted the triangles correctly :P 

Joke aside, I updated the PR to fit your suggestions: minor observation though, for "Using the example data, vertex 3 will collect the pairs (4, 1), (5, 1), (6, 1), (4, 2), (5, 2) and (6, 2). Again, you group by ID and attach these values to the receiving vertices." There is a shortcut there that I could not resist to use :) You'll see it in the code.

Finally, once we polish this example, I believe it would make a nice addition to the Gelly-"production" examples as there have been some questions regarding Triangle Count on the mailing list.

Tell me what you think :)

File: src/main/java/example/TriangleCount.java [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+	@Override
+	public String getDescription() {
+		return "Triangle Count Example";
+	}
+
+	/**
+	 * Each vertex will have the set of neighbors with id lower than its own id as value.
+	 */
+	@SuppressWarnings("serial")
+	private static final class GatherLowerNeighbors implements
```
Comment (vasia) [inline src/main/java/example/TriangleCount.java]: What this does is quite different from what I described. Here, each vertex looks at all its neighbors and chooses the ones with lower IDs. What we want is that each vertex looks at its neighbors and _generates_ a msg for each neighbor with higher ID. The result is the same -- what you've implemented will give us the correct result, but if we want to closely simulate the vertex-centric algorithm, we should do it the second way. I guess a `groupReduceOnEdges` should be enough for this.

File: src/main/java/example/TriangleCount.java [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+				}
+			}
+
+			collector.collect(new Vertex<String, HashSet<String>>(vertex.getId(), neighborsHashSet));
+		}
+	}
+
+	@SuppressWarnings("serial")
+	private static final class PropagateNeighborValues implements
+			NeighborsFunctionWithVertexValue<String, HashSet<String>, NullValue, Vertex<String, HashSet<String>>> {
```
Comment (vasia) [inline src/main/java/example/TriangleCount.java]: Same here. Each vertex should go through received messages, i.e. the elements inside its values and propagate each of the msgs to neighbors with higher ID, i.e. collect (neighborID, msg) pairs.

File: src/main/java/example/TriangleCount.java [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+		// simulate the second superstep
+		// propagate each received "message" to neighbours with higher id
+		DataSet<Vertex<String, HashSet<String>>> verticesWithPropagatedValues =
+				graphWithVertexNeighbors.groupReduceOnNeighbors(new PropagateNeighborValues(), EdgeDirection.ALL);
+
+		Graph<String, HashSet<String>, NullValue> graphWithPropagatedVertexValues = Graph.fromDataSet(verticesWithPropagatedValues,
+				edges, env);
+
+		DataSet<Vertex<String, Integer>> verticesWithNumberOfTriangles = graphWithPropagatedVertexValues.getTriplets()
```
Comment (vasia) [inline src/main/java/example/TriangleCount.java]: For the same reasons, we should avoid `getTriplets()`, which spreads the computation on the edges. Remember, we want to write this is a "vertex-centric" way :)
Each vertex has to check for triangles by comparing its msgs and its neighbors.

**Comment by andralungu** [comment]
PR updated to fit the very first set of comments. 
Hopefully it simulates vertex-centric (I don't think I can add more overhead than this :P ).

**Pull request merged**